Add MCP support for Redpanda Cloud

birdayz opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

I would like to avoid introducing a plugin for this for now, we'll be fine with the release cycle of core. We're not expecting this to change very often (and want to spare all the time required to do a plugin).

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes


* none

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/redpanda-data/redpanda?pullRequest=25802) <br/>All committers have signed the CLA.

birdayz commented on src/go/rpk/go.mod:
```diff
@@ -41,6 +42,7 @@ require (
 	github.com/pkg/errors v0.9.1
 	github.com/prometheus/client_model v0.6.1
 	github.com/prometheus/common v0.62.0
+	github.com/redpanda-data/common-go/proto v0.0.0-20250416183016-0cd4089182bd
```
to pull `*pb.mcp.go` files

this is currently on a branch, i will update before merging

birdayz commented on src/go/rpk/pkg/publicapi/controlplane.go:
```diff
@@ -27,10 +27,23 @@ import (
-	Organization  iamv1connect.OrganizationServiceClient
-	ResourceGroup controlplanev1connect.ResourceGroupServiceClient
-	Serverless    controlplanev1connect.ServerlessClusterServiceClient
+	Region           controlplanev1connect.RegionServiceClient
```
added missing clients

birdayz commented on src/go/rpk/pkg/publicapi/publicapi.go:
```diff
@@ -33,6 +33,15 @@ func newAuthInterceptor(token string) connect.UnaryInterceptorFunc {
 	}
 }
 
+func newReloadingAuthInterceptor(f func() string) connect.UnaryInterceptorFunc {
```
is a specialized interceptor, that does not use a static token. instead, it can call a function that provides the token (which can be refreshed in the background)

r-vasquez commented on src/go/rpk/pkg/cli/cloud/mcp/mcp.go:
```diff
@@ -0,0 +1,210 @@
+func newInstall(fs afero.Fs, p *config.Params) *cobra.Command {
+	var allowDelete bool
+	cmd := &cobra.Command{
+		Use:   "install [claude]",
+		Short: "Install Redpanda Cloud MCP server. Only Claude Desktop is supported at this time.",
```
```suggestion
		Use:   "install [MCP-SERVER]",
		Short: "Install Redpanda Cloud MCP server",
		Long: `Install Redpanda Cloud MCP server.
		
Only Claude Desktop is supported at this time.`
```

And if you can, more documentation on what this command does, i.e: sets a configuration file in the user configuration directory, etc...

birdayz replied:
done

r-vasquez commented on src/go/rpk/pkg/cli/cloud/mcp/mcp.go:
```diff
@@ -0,0 +1,210 @@
+			cfg, err := p.Load(fs)
+			out.MaybeDie(err, "Failed to load rpk config: %w", err)
+
+			if args[0] != "claude" {
```
Let's make this a flag so it's easier to change in the future. `--server` or something similar. Managing conflicting args is trickier in the long run.

You could also make the flag required by default [(example)](https://github.com/redpanda-data/redpanda/blob/bf6cbc1f401c2caef9c49be0ca801d7aafb9b395/src/go/rpk/pkg/cli/security/role/assign.go#L74) and add a autocomplete with the current valid servers [(example)](https://github.com/redpanda-data/redpanda/blob/bf6cbc1f401c2caef9c49be0ca801d7aafb9b395/src/go/rpk/pkg/cli/registry/mode/set.go#L88)

birdayz replied:
done, can you check if i did this correctly? 

https://github.com/redpanda-data/redpanda/blob/jb%2Fwip-mcp/src/go/rpk/pkg/cli/cloud/mcp/mcp.go#L105

r-vasquez commented on src/go/rpk/pkg/cli/cloud/mcp/mcp.go:
```diff
@@ -0,0 +1,210 @@
+				out.DieString("Argument claude is missing")
+			}
+			configDir, err := os.UserConfigDir()
+			out.MaybeDie(err, "failed to load file: %w", err)
```
[nit]
```suggestion
			out.MaybeDie(err, "failed to get user configuration directory: %w", err)
```

birdayz replied:
done

r-vasquez commented on src/go/rpk/pkg/cli/cloud/mcp/mcp.go:
```diff
@@ -0,0 +1,210 @@
+
+			// Set dev overrides
+			if cfg.DevOverrides().PublicAPIURL != "" {
+				jsonStr, err = sjson.Set(jsonStr, "mcpServers.redpandaCloud.env", map[string]any{
+					"RPK_CLOUD_URL":           "https://cloud-api.ign.cloud.redpanda.com",
+					"RPK_CLOUD_AUTH_URL":      "https://integration-cloudv2.us.auth0.com",
+					"RPK_CLOUD_AUTH_AUDIENCE": "cloudv2-ign.redpanda.cloud",
+					"CLOUD_URL":               "https://cloud-api.ign.cloud.redpanda.com/api/v1",
+					"RPK_AUTH_APP_CLIENT_ID":  "OQJLrKFCXuCMfGfMfnIqzgiwiyDfoxEV",
+					"RPK_PUBLIC_API_URL":      "https://api.ign.cloud.redpanda.com",
+				})
```
Can we use the new `RPK_CLOUD_ENVIRONMENT=integration`?

birdayz replied:
how can i programmatically access the -X cloud_environment flag?

r-vasquez replied:
It gets added on load to the profile https://github.com/redpanda-data/redpanda/blob/dev/src/go/rpk/pkg/config/rpk_yaml.go#L148

It was not intended to be documented nor part of the dev overrides as we didn't see a use for it, but feel free to explore if this suits you or if we need to make a change.

birdayz replied:
done

r-vasquez commented on src/go/rpk/pkg/cli/cloud/mcp/mcp.go:
```diff
@@ -0,0 +1,210 @@
+			out.MaybeDie(err, "failed to patch Claude Desktop config: %v", err)
+
+			// Set dev overrides
+			if cfg.DevOverrides().PublicAPIURL != "" {
```
[question] Why do we set explicitly integration if the PublicAPIURL != "" ? what if it's preprod?

Shouldn't it be

```
RPK_CLOUD_URL = cfg.DevOverrides().CloudAPIURL
```
And so on ... ?

That's if `RPK_CLOUD_ENVIRONMENT` is not enough though.

birdayz replied:
i switched everything to cloud_env via VirtualProfile now

r-vasquez commented on src/go/rpk/pkg/cli/cloud/mcp/mcp.go:
```diff
@@ -0,0 +1,210 @@
+		Use:   "stdio",
+		Short: `MCP stdio server. Can be used by MCP clients.`,
+		Args:  cobra.NoArgs,
+		Run: func(cmd *cobra.Command, args []string) {
```
```suggestion
		Run: func(_ *cobra.Command, args []string) {
```

birdayz replied:
done

r-vasquez commented on src/go/rpk/pkg/cli/cloud/mcp/mcp.go:
```diff
@@ -0,0 +1,210 @@
+					m.Lock()
+					tokenOK = false
+					m.Unlock()
+					return fmt.Errorf("failed to validate Cloud token. This may happen, if they never logged in to Redpand Cloud. However, their token is not just expired, we'd not see this error then. Instruct the user to run `rpk%s cloud login` to login to Redpanda Cloud. After running the command, they can ask to retry.", extra)
```
Is this an instruction to the LLM, or is it going straight to the end user? 

If this is going to be returned to the user, then:
```suggestion
					return fmt.Errorf("failed to validate your Redpanda Cloud token. This can happen if you have never logged in to Redpanda Cloud before. To fix this, please run 'rpk%s cloud login' to log in to Redpanda Cloud. Once you have logged in, you can retry your request", extra)
```

Feel free to ignore otherwise

birdayz replied:
It goes to the LLM. What the end user sees is up to the LLM

r-vasquez commented on src/go/rpk/pkg/cli/cloud/mcp/mcp.go:
```diff
@@ -0,0 +1,210 @@
+					m.Lock()
+					tokenOK = false
+					m.Unlock()
+					return fmt.Errorf("Redpanda Cloud token is expired. Instruct the user to run `rpk%s cloud login` to obtain a fresh one. Afterwards, they can ask to retry.", extra) //nolint:revive // We want it to start with capital letter, because Redpanda is a name
```
Same question as above. 
```suggestion
					return fmt.Errorf("Redpanda Cloud token is expired. Run `rpk%s cloud login` to obtain a fresh one and retry again", extra) //nolint:revive // We want it to start with a capital letter, because Redpanda is a name
```

birdayz replied:
It goes to the LLM. What the end user sees is up to the LLM

r-vasquez commented on src/go/rpk/pkg/cli/cloud/mcp/mcp.go:
```diff
@@ -0,0 +1,210 @@
+		},
+	}
+
+	cmd.Flags().BoolVarP(&allowDelete, "allow-delete", "", false, "Allow delete RPCs. Off by default.")
```
```suggestion
	cmd.Flags().BoolVarP(&allowDelete, "allow-delete", "", false, "Allow delete RPCs. Off by default")
```

We don't add a period in the flag help text (sorry, we don't have a linter for this yet).

birdayz replied:
done

r-vasquez commented:
There seem to be missing files for the bazel build, @birdayz please run `make bazel` to autogenerate the files.

(It's worth rebasing if you are using a Mac as we fixed a bug for the autogeneration in https://github.com/redpanda-data/redpanda/pull/25368)

birdayz committed:
rpk: add cloud MCP support

- Add command to install the MCP server. Currently, only claude desktop
  is supported.
- Add command to serve the stdio server. MCP clients launch it as a
  subprocess, and interact over stdin/stdout with it.
- Authentication to cloud is completely re-used from rpk.

birdayz commented:
i addressed the comments, could you have a look once more @r-vasquez ? thanks!

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--698e6a2e-7441-44f5-a828-6d4598cf1154-->
<!--TEST RESULTS-->
<!--119237c5-0002-42d4-be43-8d19b40deaa9-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64852">build#64852</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_raft_rpunit.gtest_raft_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/64852#01965d2b-3106-448c-ad5a-4ca68873945d | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/64852#01965d86-13c8-4c77-80cb-cf5592ef0ba2 | FLAKY | 16/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64873">build#64873</a></summary>


| test_id                                                                                              | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-----------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_raft_rpunit.gtest_raft_rpunit                                                                  | unit        | https://buildkite.com/redpanda/redpanda/builds/64873#01965e88-6938-4f33-85a6-53db284ce0de | FLAKY         | 1/2      |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64873#01965ee1-2e69-4dca-b664-e7c29d01e36a | FLAKY         | 16/21    |
</details>
<!--/body-->

r-vasquez commented on src/go/rpk/pkg/cli/cloud/mcp/mcp.go:
```diff
@@ -0,0 +1,206 @@
+	var allowDelete bool
+	var mcpClient string
+	cmd := &cobra.Command{
+		Use:   "install [MCP-SERVER]",
```
```suggestion
		Use:   "install",
```

birdayz replied:
updated

r-vasquez commented on src/go/rpk/pkg/cli/cloud/mcp/mcp.go:
```diff
@@ -0,0 +1,206 @@
+			out.MaybeDie(err, "failed to rename: %v", err)
+		},
+	}
+	cmd.Flags().BoolVarP(&allowDelete, "allow-delete", "", false, "Allow delete RPCs. Off by default")
+	cmd.Flags().StringVar(&mcpClient, "client", "claude", "Name of the MCP client to configure. Default: claude")
```
```suggestion
	cmd.Flags().BoolVarP(&allowDelete, "allow-delete", "", false, "Allow delete RPCs")
	cmd.Flags().StringVar(&mcpClient, "client", "claude", "Name of the MCP client to configure")
```
Cobra already prints the default.

nits, feel free to ignore.

birdayz replied:
updated

r-vasquez commented on src/go/rpk/pkg/cli/cloud/mcp/mcp.go:
```diff
@@ -0,0 +1,206 @@
+
+func newInstall(fs afero.Fs, p *config.Params) *cobra.Command {
+	var allowDelete bool
+	var mcpClient string
```
Are you using or validating this flag in the run function?

It is enforced, but the command assumes it is Claude.

birdayz replied:
updated

r-vasquez reviewed (dismissed):
LGTM, thanks!

https://github.com/redpanda-data/redpanda/pull/25802/files#r2048801329 <- Let me know when you update this and I'll approve right away.

vbotbuildovich commented:
/backport v25.1.x
